### ISSUE TITLE
Improve the newtype wrapper situation

### ISSUE THREAD

**Issue by dtolnay** [serde-rs/serde#496]
Something like:

``` rust
serde_newtype! {
    #[derive(Serialize, Deserialize)]
    pub struct MyDuration(time::Duration);

    struct time::Duration {
        secs: u64,
        nanos: u32,
    }
}
```

**Comment by dtolnay** [comment]
@nox points out the approach from hyper_serde could be more convenient in some cases so we can provide that instead / in addition to the above. That approach is to define a single shared `Ser<T>` and `De<T>` wrapper in the user's crate and then implement Serialize for `Ser<time::Duration>` and Deserialize for `De<time::Duration>`. Then provide functions to use with serialize_with and deserialize_with that take any `Ser<T>` or `De<T>`. That lets you use the intended type like time::Duration inside all of your structs rather than using a newtype wrapper inside your structs.

**Comment by dtolnay** [comment]
That could look like:

```rust
// just once
serde_remote_init!(SerdeWrapper);

// for each struct
serde_remote_derive! {
    #[derive(Serialize, Deserialize)]
    struct time::Duration {
        secs: u64,
        nanos: u32,
    }
}

#[derive(Serialize, Deserialize)]
struct Mine {
    #[serde(serialize_with = "SerdeWrapper::serialize",
            deserialize_with = "SerdeWrapper::deserialize")]
    dur: time::Duration,
}
```

**Comment by dtolnay** [comment]
I wonder if I can make this work:

```rust
#[derive(Serialize, Deserialize)]
#[serde(remote = "time::Duration")]
struct Duration {
    secs: u64,
    nanos: u32,
}

#[derive(Serialize, Deserialize)]
struct Mine {
    #[serde(with = "Duration")]
    dur: time::Duration,
}
```

**Comment by dtolnay** [comment]
Private fields make it less nice but here is a proof of concept:

```rust
#[macro_use]
extern crate serde_derive;

extern crate serde;
extern crate serde_json;

use serde::ser::{Serializer, SerializeStruct};

use std::time;

// #[derive(Serialize)]
// #[serde(remote = "time::Duration")]
struct Duration {
    // #[serde(getter = "time::Duration::as_secs")]
    secs: u64,
    // #[serde(getter = "time::Duration::subsec_nanos")]
    nanos: u32,
}

impl Duration {
    fn serialize<S>(remote: &time::Duration, serializer: S) -> Result<S::Ok, S::Error>
        where S: Serializer
    {
        let mut state = serializer.serialize_struct("Duration", 2)?;
        state.serialize_field("secs", &time::Duration::as_secs(&remote))?;
        state.serialize_field("nanos", &time::Duration::subsec_nanos(&remote))?;
        state.end()
    }
}

#[derive(Serialize)]
struct Mine {
    #[serde(with = "Duration")]
    dur: time::Duration,
}

fn main() {
    let mine = Mine { dur: time::Duration::new(1, 2) };
    println!("{}", serde_json::to_string(&mine).unwrap());
}
```

**Comment by dtolnay** [comment]
#### Proposed semantics

If the remote struct has all public fields then it works with just a remote attribute.

```rust
#[derive(Serialize, Deserialize)]
#[serde(remote = "time::Duration")]
struct Duration {
    secs: u64,
    nanos: u32,
}
```

If the remote struct has one or more private fields, you provide a getter for any private fields and an Into conversion in the deserialization direction. The Into conversion is executed if there is at least one getter attribute.

```rust
#[derive(Serialize, Deserialize)]
#[serde(remote = "time::Duration")]
struct Duration {
    #[serde(getter = "time::Duration::as_secs")]
    secs: u64,
    #[serde(getter = "time::Duration::subsec_nanos")]
    nanos: u32,
}

impl From<Duration> for time::Duration {
    fn from(helper: Duration) -> time::Duration {
        time::Duration::new(helper.secs, helper.nanos)
    }
}
```

**Comment by dtolnay** [comment]
Fixed in #858.
